pytorch_lightning/callbacks/base.py

```diff
@@ -178,3 +178,19 @@ def on_before_zero_grad(self, trainer, pl_module, optimizer):
         Called after optimizer.step() and before optimizer.zero_grad().
         """
         pass
+
+    def log(self, *_, **__):
+        """
+        This function is a placeholder. Don't override this function.
+        When using Lightning Trainer, the log function from
+        the model will be attached to this callback.
+        """
+        pass
```

Comment:
I still don't understand why this is just a placeholder. Why can't we make the logging call from here? Why is it necessary to delay and dynamically set the method? 
If this was just a regular method you wouldn't need to do all the strange warning messages and checks for overriding.

Reply:
well, we probably could as each callback has a trainer, and each trainer has a ref to model... we can just raise an error if the user tries to use it before the model is linked...

Reply:
Yes, something like that is what I mean exactly.

Reply:
Hey @awaelchli,

Users can currently do `pl_module.log(...)` in their callback functions, but it makes sense to be able to do `self.log` everywhere.

The log function is tied to the model. It will save the logged metrics in the `model.results = Result()` object, and then will save its reference to EpochResultStore.    

To enable what you want, we would need to move log function to trainer for both model and callbacks to call it internally, move results object to the trainer, add trainer reference to each callbacks, still make sure people don't override log and log_dict. 
It is possible by involve a slightly bigger refactor than attaching the log function to callbacks.

If you feel it isn't right, let's merge this PR as log functions are already attached and work on a refactor.

Best regards,
T.C

Reply:
We will do this in another PR. We will need to move logging to the Trainer, so it is cleaner.

Reply:
I would rather do it in this one as this seems tobe the main reason for this PR :]